docs/SAW.rst

```diff
@@ -2,9 +2,419 @@
+=============
+
+SAW allows one to direct a verification task using a proof script, which is simply a sequence of proof
+tactics to apply. Very commonly, the proof script provided in a verification task is simply an instruction
+to use an external SAT/SMT solver such as ABS, Yices, or Z3.
+
+A proof script is represented as a JSON object with a single field:
+
+``tactics``
+  A list of proof tactics to apply to the context/goal. A proof tactic is represented as a JSON object
+  containing a tag named ``tactic``, with any further fields determined by this tag. These tag values can be:
+
+  ``use prover``
+    Apply an external prover to the goal. There is an additional field ``prover`` which is a JSON object
+    with a field ``name`` specifying what prover to use (one of ``abc``, ``cvc4``, ``rme``, ``yices``, or ``z3``),
+    and a field ``uninterpreted functions`` when ``name`` is one of ``cvc4``, ``yices``, or ``z3``. This
+    field is a list of names of functions taken as uninterpreted/abstract.
+
+  ``unfold``
+    Unfold terms in the context/goal. There is an additional field ``names``, a list of the names bound on
+    the server to unfold.
+
+  ``beta reduce goal``
+    Perform a single beta reduction on the proof goal.
+
+  ``evaluate goal``
+    Fully evaluate the proof goal. There is an additional field ``uninterpreted functions``, a list of names
+    of functions taken as uninterpreted/abstract.
+
+  ``simplify``
+    Simplify the context/goal. There is an additional field ``rules``, a name bound to a simpset on the server.
+
+  ``assume unsat``
+    Assume the goal is unsatisfiable, which in the current implementation of SAW should be interpreted as
+    assuming the property being checked to be true. This is likely to change in the future.
+
+  ``trivial``
+    States that the goal should be trivially true (either the constant ``True`` or a function that immediately
+    returns ``True``. This tactic fails if that is not the case.
+
+.. _setup-values:
+
+Crucible Setup Values
+=====================
+
+Setup Values encompass all values that can occur during symbolic execution, including Cryptol terms,
+pointers, arrays, and structures. They are used extensively when writing the specifications provided to the
+``verify`` commads. Setup Values are represented as JSON objects containing a tag field, ``setup value``,
+that determines the other fields. This tag value can be:
+
+``saved``
+  A term previously saved on the server. There is an additional field ``name`` giving the name bound to the
+  term on the server.
+
+``null value``
+  A null/empty value.
+
+``Cryptol``
+  A Cryptol term. There is an additional field ``expression`` containing a Cryptol expression.
+
+``array value``
+  An array value. There is an additional field ``elements`` which is a list of :ref:`Crucible Setup values<setup-values>`
+  to populate the array with.
+
+``field lvalue``
+  A struct field to be assigned to. There are two additional fields:
```

Comment:
Reading this makes me realize that `field lvalue` and `element lvalue` aren't _only_ for destinations. They should perhaps have a different name.

Reply:
I didn't do anything with this one just yet - Perhaps I don't have a broad enough view of what an lvalue is, and that's more the problem than the name? :)

Reply:
I should look at the code again and decide whether a better description is warranted.

Reply:
Sounds good!

Reply:
So my own digging in to this suggests that you're right; these can be used as more than the targets of assignments. This is mostly an educated guess, though, based on naming and my (admittedly limited) understanding of the internals of Crucible :)

Reply:
Okay, cool. Then perhaps just "A field of a struct" instead of "A struct field to be assigned to" would be more appropriate. I think we can leave `lvalue` as part of the name, since that would require changing some code, though we could do that if it became confusing.